File: integrations/langgraph/typescript/src/interrupts.ts
import { randomUUID } from "@ag-ui/client";
import type { Interrupt as LangGraphInterrupt } from "@langchain/langgraph-sdk";
import type { Interrupt as AGUIInterrupt, ResumeEntry } from "@ag-ui/core";

const isPlainObject = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null && !Array.isArray(v);

export function langGraphInterruptToAGUI(
  lg: LangGraphInterrupt,
): AGUIInterrupt {
  const raw = lg.value;
  const dict = isPlainObject(raw) ? raw : null;

  const id = lg.id || `lg-${randomUUID()}`;
  const reason =
    (dict?.reason as string | undefined) ?? "langgraph:interrupt";

  const message =
    typeof raw === "string"
      ? raw
      : (dict?.message as string | undefined);
  const toolCallId =
    (dict?.toolCallId as string | undefined) ??
    (dict?.tool_call_id as string | undefined);
  const responseSchema =
    (dict?.responseSchema as Record<string, unknown> | undefined) ??
    (dict?.response_schema as Record<string, unknown> | undefined);
  const expiresAt =
    (dict?.expiresAt as string | undefined) ??
    (dict?.expires_at as string | undefined);

  const metadata: Record<string, unknown> = {
    langgraph: {
      raw,
      ns: (lg as { ns?: string[] }).ns,
      resumable: (lg as { resumable?: boolean }).resumable,
      when: (lg as { when?: string }).when,
    },
  };

  return {
    id,
    reason,
    ...(message !== undefined ? { message } : {}),
    ...(toolCallId !== undefined ? { toolCallId } : {}),
    ...(responseSchema !== undefined ? { responseSchema } : {}),
    ...(expiresAt !== undefined ? { expiresAt } : {}),
    metadata,
  };
}

export function langGraphInterruptsToAGUI(
  list: readonly LangGraphInterrupt[],
): AGUIInterrupt[] {
  return list.map(langGraphInterruptToAGUI);
}

export const DEFAULT_RESUME_SENTINEL_CANCELLED = "__agui_cancelled__";
export const DEFAULT_RESUME_SENTINEL_MAP = "__agui_resume_map__";

export function buildLgCommandResumeFromAgui(
  entries: readonly ResumeEntry[],
): unknown {
  if (entries.length === 1) {
    const e = entries[0];
    if (e.status === "resolved") return e.payload;
    return { [DEFAULT_RESUME_SENTINEL_CANCELLED]: true, interrupt_id: e.interruptId };
  }
  return {
    [DEFAULT_RESUME_SENTINEL_MAP]: Object.fromEntries(
      entries.map((e) => [
        e.interruptId,
        { status: e.status, payload: e.payload ?? null },
      ]),
    ),
  };
}

File: integrations/langgraph/python/ag_ui_langgraph/interrupts.py
from typing import Any, List, Mapping
from uuid import uuid4

from ag_ui.core import Interrupt as AGUIInterrupt
from langgraph.types import Interrupt as LangGraphInterrupt

from .utils import make_json_safe


def lg_interrupt_to_agui(lg: LangGraphInterrupt) -> AGUIInterrupt:
    raw = lg.value
    is_dict = isinstance(raw, Mapping)

    interrupt_id = getattr(lg, "id", None) or f"lg-{uuid4()}"
    reason = (raw.get("reason") if is_dict else None) or "langgraph:interrupt"

    message = (
        raw if isinstance(raw, str)
        else raw.get("message") if is_dict else None
    )
    tool_call_id = (
        raw.get("toolCallId") if is_dict else None
    ) or (
        raw.get("tool_call_id") if is_dict else None
    )
    response_schema = (
        raw.get("responseSchema") if is_dict else None
    ) or (
        raw.get("response_schema") if is_dict else None
    )
    expires_at = (
        raw.get("expiresAt") if is_dict else None
    ) or (
        raw.get("expires_at") if is_dict else None
    )

    metadata: dict[str, Any] = {
        "langgraph": {
            "raw": make_json_safe(raw),
            "ns": getattr(lg, "ns", None),
            "resumable": getattr(lg, "resumable", None),
            "when": getattr(lg, "when", None),
        }
    }

    return AGUIInterrupt(
        id=interrupt_id,
        reason=reason,
        message=message,
        tool_call_id=tool_call_id,
        response_schema=response_schema,
        expires_at=expires_at,
        metadata=metadata,
    )


def lg_interrupts_to_agui(items) -> List[AGUIInterrupt]:
    return [lg_interrupt_to_agui(i) for i in items]


DEFAULT_RESUME_SENTINEL_CANCELLED = "__agui_cancelled__"
DEFAULT_RESUME_SENTINEL_MAP = "__agui_resume_map__"
